### ISSUE TITLE
Flux sampling issues

### ISSUE THREAD

**Issue by h8dinh** [opencobra/cobrapy#328]
Hi,

I have tried the new cobrapy features cobra.flux_analysis.sampling (flux sampling, used sampling.py file, version in branch devel2 on Wednesday, December 7). There are two issues I have seen when running the simulation with a modified E. coli genome scale model iJO1366, glucose uptake -20, anaerobic. I tries to run simulation for the same case two times to compare the two output distribution.

1) First issue, I tried to see if I use the sampling method two times, is there any different between the two cases? Using method='optgp', processes=4, sample=1000 or 20000 yield exact same distribution, which worries me. However, using 'arch' sample=1000 yield different distribution.
(e.g. code: sampling.sample(model, 20000, processes=4, method='optgp', solver='gurobi'))

2) Second issue, when I tried to use 'arch' sample=10000 or 20000 two times, the first time it runs through ok, but the second time it yields an error:
RuntimeError: maximum recursion depth exceeded
(e.g. code: sampling.sample(model, 10000, method='arch', solver='gurobi'))

Additional info:
- I implement the method by simply copying the sampling.py python file to my local updated cobrapy master branch. Then, I just use sampling module by import cobra.flux_analysis.sampling as sampling.

**Comment by h8dinh** [comment]
For the second issue, when I delete the object containing the distribution before assigning the sampling output to that object, somehow the error of recursion depth go away (not sure if this work around is replicable).

E.g.
del wt
(simulation)
wt = sampling.sample(model, 10000, method='arch', solver='gurobi')

**Comment by h8dinh** [comment]
For the first issue, there is something wrong with setting multiple core. When I run optgp with 1 core, everything is ok. When I run with 4 cores, I got two exact same distribution.

For sample=1000, method='optgp', solver='gurobi'
1 core:
![image](https://cloud.githubusercontent.com/assets/15752681/21121471/772a7bb8-c081-11e6-94fb-7565af328375.png)


4 cores:
![image](https://cloud.githubusercontent.com/assets/15752681/21121468/6b3189d2-c081-11e6-950e-de622b12d352.png)

**Comment by cdiener** [comment]
Hi, thanks for the testing! I pretty much know what's up with the second option (has to do how we reset the step when sampling produces bound violations). The first one is interesting though. It would help if you could give me the following information:

1. Your Operating System (Windows, Mac, Linux)
2. Your Python version

Also it would be great if you could lend me your model. For instance if you have your model already in code and it is named `model` you could do

```python
from cobra.io import save_json_model
save_json_model(model, "model.json")
```

If you zip it you can attach it directly to your reply.

Thanks in advance!

**Comment by h8dinh** [comment]
Hi, thanks for your reply.

My operating system is Linux (Ubuntu 16.04 LTS). My python version is 2.7.11. The model attached can be loaded using pickle (let me know if you really need the json format one).

Here's the code I used:
with open('./input_data/iJO1366_modABEIJK.pickle') as f:
    model = cloudpickle.load(f)
model.reactions.EX_glc__D_e.lower_bound = -20.
model.reactions.BIOMASS_Ec_iJO1366_core_53p95M.objective_coefficient = 1.
fbag = model.optimize(solver='gurobi')

model.reactions.BIOMASS_Ec_iJO1366_core_53p95M.lower_bound = 0.99*fbag.f
swt1 = sampling.sample(model, 1000, method='optgp', processes=4, solver='gurobi')

with open('./input_data/iJO1366_modABEIJK.pickle') as f:
    model = cloudpickle.load(f)
model.reactions.EX_glc__D_e.lower_bound = -20.
model.reactions.BIOMASS_Ec_iJO1366_core_53p95M.objective_coefficient = 1.
fbag = model.optimize(solver='gurobi')

model.reactions.BIOMASS_Ec_iJO1366_core_53p95M.lower_bound = 0.99*fbag.f
swt2 = sampling.sample(model, 1000, method='optgp', processes=4, solver='gurobi')

plt.figure(1, figsize=(10,10))
for i,rxn in enumerate(['EX_glc__D_e', 'EX_etoh_e', 'EX_ac_e', 'EX_lac__D_e', 'GAPD', 'G6PDH2r']):
    plt.subplot(330+i+1)
    a = swt1[:,rnum[rxn]]
    b = swt2[:,rnum[rxn]]
    bins = np.histogram(np.hstack((a,b)), bins=50)[1]
    plt.hist(a, bins, color='red', alpha=0.5)
    plt.hist(b, bins, color='blue', alpha=0.5)
    plt.title(rxn)

plt.subplot(333)
plt.legend(bbox_to_anchor=(2.3,0.48), numpoints=1, loc='lower right')

[model.zip](https://github.com/opencobra/cobrapy/files/650059/model.zip)

**Comment by cdiener** [comment]
Thanks. I think I already fixed the issues, however, I will continue to test with your model. In the meantime you can try the `sampling.py` [from here](https://raw.githubusercontent.com/cdiener/cobrapy/fix/sampling/cobra/flux_analysis/sampling.py).

Cheers

**Comment by cdiener** [comment]
Closing for now since the fixes have been merged with #329. Feel free to reopen if the problems persist :smiley:
